Zipkin Exporter: Specify RemoteEndpoint

CodeBlanch opened the pull request:
The main goal of this PR was to pass RemoteEndpoint data to the Zipkin API so that it can identify spans that are actually flowing to external services, but might not be instrumented (3rd party). This allows it to correctly treat them as dependencies.

Somewhat fuzzy logic to accomplish its goal. If "net.peer.name" or "peer.service" is set, it will use it. Otherwise it will attempt "peer.hostname," "peer.address," "http.host," and "db.instance" in that order. Should work out of the box for the existing OpenTelemetry collectors (and the Sql one in PR).

Work also done...

* Added caching of ZipkinEndpoints. There was a "todo" in the code before.
* Switched to System.Text.Json for .NET Standard 2.0 target. That was done for speed. No problem removing it if there are objections.

MikeGoldsmith commented on src/OpenTelemetry.Exporter.Zipkin/ZipkinTraceExporter.cs:
```diff
@@ -131,20 +145,51 @@ internal ZipkinSpan GenerateSpan(SpanData otelSpan, ZipkinEndpoint defaultLocalE
                 spanBuilder.ParentId(this.EncodeSpanId(otelSpan.ParentSpanId));
             }
 
+            string peerService = null; // RemoteEndpoint.ServiceName primary.
```
This could be simplified to just be a single string and only overwrite if null - essentially moving the logic on line 229 to maintain a single value.

CodeBlanch replied:
@MikeGoldsmith That logic would tweak the behavior slightly in that, depending on the order of the attributes and the content, the last match would win. I updated with a more efficient solution that does what you were suggesting in spirit but also preserves the "priority" that was in there. See below.

MikeGoldsmith commented on src/OpenTelemetry.Exporter.Zipkin/ZipkinTraceExporter.cs:
```diff
@@ -131,20 +145,51 @@ internal ZipkinSpan GenerateSpan(SpanData otelSpan, ZipkinEndpoint defaultLocalE
+
+                if (strVal != null)
+                {
+                    if (key == "peer.service")
```
A switch statement would be more concise.

CodeBlanch replied:
@MikeGoldsmith Went with a case-insensitive dictionary.

CodeBlanch commented:
@lmolkova I added support for "net.peer.name" based on OTel attribute list you pointed out, thanks.

@MikeGoldsmith 

* I moved the "if" statements into a case-insensitive dictionary and added in the concept of a priority. What I didn't want was a scenario where it would be selecting "db.instance" over "net.peer.name" should both be supplied just because "db.instance" happened to come later. The logic became an order of magnitude more complex so I added a unit test project and refactored `GenerateSpan` into `ToZipkinSpan` so it could be tested nicely. I modeled after the Jeager Exporter.
* The project targets net46 & netstandard2.0. System.Text.Json support .NET 4.6.1 and up. If we bump the version we could eliminate the need for JSON.Net?

Regarding waiting on the spec to formalize, works for me! Just wanted to get all the feedback addressed while it was fresh.

MikeGoldsmith commented:
Hey @CodeBlanch 

We've discussed target framework version a few times, and the general consensus is that we should target the lowest available version for API and SDK. Exporters & Collectors should aim for as low as possible, but don't have the same restriction. eg the StackExchangeRedisCollector targets net462 & netstandard2.0.

MikeGoldsmith commented on src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinConversionExtensions.cs:
```diff
@@ -0,0 +1,187 @@
+
+                if (strVal != null
+                    && RemoteEndpointServiceNameKeyResolutionDictionary.TryGetValue(key, out int priority)
+                    && (remoteEndpointServiceName == null || remoteEndpointServiceName.Item2 > priority))
```
I think this should be `<` as the priority is in ascending order from 0 - we want the lowest ranked peer.service.

CodeBlanch replied:
It actually works, there's a unit test for it. But something about the way it is written is confusing. I flipped it around I think it reads better now. LMK

CodeBlanch commented:
For the json library part of this, here's some performance analysis...

net462 w/ Json.NET:
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4121.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4121.0), X64 RyuJIT


```
|                     Method | NumberOfSpans |     Mean |     Error |    StdDev |    Median |   Gen 0 |   Gen 1 |  Gen 2 | Allocated |
|--------------------------- |-------------- |---------:|----------:|----------:|----------:|--------:|--------:|-------:|----------:|
| **ZipkinExporter_ExportAsync** |             **1** | **1.082 ms** | **0.0719 ms** | **0.1931 ms** | **0.9799 ms** |  **7.8125** |  **1.9531** |      **-** |  **49.27 KB** |
| **ZipkinExporter_ExportAsync** |            **50** | **2.606 ms** | **0.1985 ms** | **0.5853 ms** | **2.6461 ms** | **42.9688** | **42.9688** | **7.8125** | **280.17 KB** |
| **ZipkinExporter_ExportAsync** |           **100** | **2.622 ms** | **0.2588 ms** | **0.7591 ms** | **2.5529 ms** | **62.5000** | **62.5000** |      **-** | **503.09 KB** |


netcoreapp31 w/ System.Text.Json:
``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET Core SDK=3.1.200-preview-014883
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|                     Method | NumberOfSpans |     Mean |     Error |    StdDev |   Gen 0 |  Gen 1 | Gen 2 | Allocated | Completed Work Items | Lock Contentions |
|--------------------------- |-------------- |---------:|----------:|----------:|--------:|-------:|------:|----------:|---------------------:|-----------------:|
| **ZipkinExporter_ExportAsync** |             **1** | **1.590 ms** | **0.0308 ms** | **0.0316 ms** |  **3.9063** | **1.9531** |     **-** |  **29.22 KB** |               **4.0039** |                **-** |
| **ZipkinExporter_ExportAsync** |            **50** | **1.845 ms** | **0.0200 ms** | **0.0178 ms** |  **9.7656** | **3.9063** |     **-** |  **86.14 KB** |               **3.0098** |                **-** |
| **ZipkinExporter_ExportAsync** |           **100** | **2.190 ms** | **0.0353 ms** | **0.0331 ms** | **15.6250** | **7.8125** |     **-** | **144.84 KB** |               **3.0078** |                **-** |

MikeGoldsmith commented:
I'd still prefer to see the exporter consistently use one JSON package over switching between them. I'd be okay bumping the target framework version to net461 to support System.Text.Json. We can always re-introduce NewtonSoft either with the #if directives or as a separate Exporter.

Thanks fo the clarification on priority - I missed the unit test but also agree the updated comparison looks cleaner. I think the code changes look good but still think we should leave this 'on-hold' until [Transformation to Zipkin]9https://github.com/open-telemetry/opentelemetry-specification/pull/380) is merged.

CodeBlanch commented:
@MikeGoldsmith No problem, updated to target just netstandard2.0 (works with net461) and removed the directives for Json.NET.

SergeyKanzhelev commented:
@lmolkova any more concerns?

SergeyKanzhelev commented:
@CodeBlanch would you like to send a PR to specs similar to https://github.com/open-telemetry/opentelemetry-specification/pull/472?

CodeBlanch commented:
@SergeyKanzhelev Sure NP, PR in the spec repo opened.
